### PR TITLE
Proptest take 2

### PR DESCRIPTION
**Pull request by timotree3**
This PR adds a test with randomly generated inputs in an attempt to find bugs in the DNA.

Based on #110

This test should just run with `make test`. If you want to run the proptest repeatedly or you want to see the debug output, and you've already built the DNA, then you can run

```
cargo test batching -- --nocapture
```

### CODE REVIEW

File: zomes/chat/src/batching_helper.rs [hunk 1]
```diff
@@ -42,44 +42,54 @@ pub fn get_message_links(
...
+                current_search_path,
+                depth,
+                children
+                    .iter()
+                    .map(|l| (&l.tag, l.timestamp))
+                    .collect::<Vec<_>>()
+            );
+            let mut children = children
+                .into_iter()
+                .filter_map(|l| path_component_from_link(&l).ok().map(|c| (c, l))) // filter out non-path links
```
Comment (timotree3) [inline zomes/chat/src/batching_helper.rs:60]: Once you get to the channel level where each child represents a year, children contains both `ChannelInfo` links and path links. We only want the path links, so I'm filtering them here with `filter_map`

File: zomes/chat/src/batching_helper.rs [hunk 2]
```diff
@@ -42,44 +42,54 @@ pub fn get_message_links(
...
+                .map(|(c, l)| Ok((segment_from_component(&c)?, l)))
+                .collect::<Result<Vec<_>, ChatError>>()?;
+
+            children.retain(|(segment, _)| *segment < earliest_seen_child_segment);
```
Comment (timotree3) [inline zomes/chat/src/batching_helper.rs:64]: I believe this `retain` is important, so I added it back. Explanation https://chat.holochain.org/holo/pl/acx5ey1c6tfkmy7pdodnz116aa

File: zomes/chat/src/batching_helper.rs [hunk 3]
```diff
@@ -42,44 +42,54 @@ pub fn get_message_links(
...
 fn append_message_links_recursive(
-    mut children: Vec<Link>,
+    mut children: Vec<(i32, Link)>,
     links: &mut Vec<Link>,
     target_count: usize,
     depth: u8,
 ) -> ChatResult<()> {
-    children.sort_unstable_by_key(|grandchild_link| cmp::Reverse(grandchild_link.timestamp));
-    for child_link in children {
+    children.sort_unstable_by_key(|(segment, _)| cmp::Reverse(*segment));
```
Comment (timotree3) [inline zomes/chat/src/batching_helper.rs]: It was important to sort by segment instead of timestamp, since in the new test, fake messages are inserted with chosen time-path-segments, but the timestamp is not fake and still represents the system time

File: zomes/chat/src/batching_helper.rs [hunk 4]
```diff
@@ -111,10 +123,10 @@ pub fn timestamp_into_path(path: Path, time: Timestamp) -> ChatResult<Path> {
     let time = DateTime::<Utc>::from_utc(NaiveDateTime::from_timestamp(ms, ns), Utc);
     let mut components: Vec<_> = path.into();
 
-    components.push(i64::from(time.year()).to_be_bytes().to_vec().into());
-    components.push(i64::from(time.month()).to_be_bytes().to_vec().into());
-    components.push(i64::from(time.day()).to_be_bytes().to_vec().into());
+    components.push((time.year() as i32).to_be_bytes().to_vec().into());
```
Comment (timotree3) [inline zomes/chat/src/batching_helper.rs:136]: Switched to i32 as an optimization. i64 was just wasting space

File: zomes/chat/src/entries/channel/handlers.rs [hunk 5]
```diff
@@ -36,7 +36,7 @@ pub(crate) fn create_channel(channel_input: ChannelInput) -> ChatResult<ChannelD
     create_link(
         path.path_entry_hash()?,
         info_hash,
-        HdkLinkType::Paths,
+        HdkLinkType::Any,
```
Comment (timotree3) [inline zomes/chat/src/entries/channel/handlers.rs:39]: This is that link at the channel level I was talking about which conflicts with path links. I tried switching the link type so that `path.children()` wouldn't return it, but it didn't help. You could revert this change. I'm not sure what HdkLinkType means but I'm guessing that `Any` is the correct choice here anyway

File: zomes/chat/tests/batching.rs [hunk 6]
```diff
@@ -1,98 +1,290 @@
...
+                    timestamp: Timestamp::from(chrono::Utc.ymd(2022, 1, 1).and_hms(1, 0, 0)),
+                },
+            ],
+            earliest_seen: Timestamp::from(chrono::Utc.ymd(2022, 2, 1).and_hms(0, 0, 0)),
+            target_message_count: 1
+        }),
+        vec!["0".to_owned()]
+    );
+
+    // TODO: The real implementation has a bug here
```
Comment (timotree3) [inline zomes/chat/tests/batching.rs:278]: This is one lead: the DNA in this branch returns `["13"]` but it should return `[]` given these inputs. I didn't investigate because in our real use-cases we always uses `target_message_count: 20`, but it's possible that investigating this would reveal the deeper bug with the chunking code.

**Review by zo-el — APPROVED** [review]
